File: src/mp/python/ShMemAcquirer.py
import socket
import json
from threading import *
from Queue import *
import thread

# Instance fields: 
# m_serversock		-- Server socket used by the acquiring server thread.
# m_port		-- Port the server thread binds to. 
# m_recvd_objs		-- Dictionary of concurrent queues we've received from. 
class ShMemAcquirer:    

    # Class constructor, all we do is specify the port to which it should bind. 
    def __init__(self, port, index, receive_queues):
        self.m_serversock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.m_port = port
        self.m_recvd_objs = receive_queues
        
        print self.m_port
        # Create a new thread to run the server. 
        thread.start_new_thread(ShMemAcquirer.run, (self,))
    
    # The actual server logic. 
    @staticmethod
    def run(me):        

        # Bind to the port and tell it we're ready to start listening to clients
        me.m_serversock.bind(("localhost", me.m_port))
        me.m_serversock.listen(5)
        
        # Server loop: listen for a client connection, pull out a set of bytes
        # and keep that serialized representation in a dictionary. 
        while 1:
            (clientsocket, address) = me.m_serversock.accept()
            msg = ''            

            # Receive the message from the client. 
            while 1:
                chunk = clientsocket.recv(4096)
            
                # We've received the entire message. 
                if len(chunk) == 0:
                    break
                else:
                    msg += chunk

            # Close the resource. 
            clientsocket.close()
            # Do some deserialization. 
            recvd_obj = json.loads(msg, 'utf-8')
            msg = recvd_obj['argument']
            sender = recvd_obj['releaser']  
 
            # Put the recieved state into the right receive queue. 
            me.m_recvd_objs[sender].put(msg, True)

File: src/java/mp/ShMemAcquirer.java
package mp;


import org.json.simple.*;
import org.json.simple.parser.*;

import java.io.*;
import java.net.*;
import java.util.concurrent.*;
import java.util.*;


import org.apache.commons.lang3.tuple.*;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.JsonParseException;
import org.codehaus.jackson.JsonProcessingException;
import org.codehaus.jackson.map.JsonMappingException;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.node.ObjectNode;


//
// This should be run on the node on which we want to run forked computations. 
// TODO: We currently ship *every* version of state on a fork. We could potentially keep
//       that state around and only send the latest version of state from the parent. 
//
public class ShMemAcquirer implements Runnable {
	
	
	// Server object that listens on a port for client connections. 
	private ServerSocket m_server;
	private Thread m_thread;
	private ObjectMapper m_mapper;
	
	// Put the state received from remote processes here. 
	private DeltaStore m_received_state;
	
	public ShMemAcquirer(int id, DeltaStore received_state) {
    	try {
    		Pair<String, Integer> my_address = ShMem.addresses.get(id);
    		m_server = new ServerSocket(my_address.getRight());
    		m_received_state = received_state;
    		m_mapper = new ObjectMapper();
    		start();
    	}
    	catch (IOException e) {
    		System.out.println("ShMemNode: Couldn't listen on port!!!");
    	    System.exit(-1);
    	}
	}
	
	public void start() {
		m_thread = new Thread(this);
		m_thread.start();
	}
	
	@SuppressWarnings("unchecked")
	public void run() {
		while (true) {
			
			// State associated with a client connection. 
			ObjectNode value = null;
			int sender = -1;
			
			// Wait for a client to connect and initialize the input and output streams.
			try {
				
				// Accept a connection from some client and grab the input stream corresponding to the connection.
				Socket client = m_server.accept();
				InputStream is = client.getInputStream();
				
				// Parse the JSON, get appropriate values of the diff and the client who released. 
				ObjectNode wrapper = (ObjectNode)m_mapper.readTree(is);
				value = (ObjectNode)wrapper.get("argument");
				sender = wrapper.get("releaser").getIntValue();
				
				// Close the input stream and client socket. 
				is.close();
				client.close();
			}
			catch (JsonProcessingException e) {
				System.err.println("Coudln't parse JSON!");
				System.exit(-1);
			}
			// If we see an error, it's probably our fault. Quit immediately. 
			catch(IOException e) {
				e.printStackTrace();
				System.exit(-1);
			}
			
			m_received_state.Push(sender,  value);
		}
	}
	
	@SuppressWarnings("unchecked")
	private JSONObject failure_message() {
		JSONObject ret = new JSONObject();
		ret.put("failure",  1);
		return ret;
	}
}